src/block_ffm.rs

```diff
@@ -225,149 +127,166 @@ impl<L: OptimizerTrait + 'static> BlockTrait for BlockFFM<L> {
-                                            zfc += fc;
-                                            addr += FFMK as usize
-                                        }
+                            continue;
+                        }
+
+                        let mut feature_num = 0;
+                        while ffm_buffer_index < fb.ffm_buffer.len() && fb.ffm_buffer.get_unchecked(ffm_buffer_index).contra_field_index == field_index_ffmk {
+                            _mm_prefetch(mem::transmute::<&f32, &i8>(&ffm_weights.get_unchecked(fb.ffm_buffer.get_unchecked(ffm_buffer_index + 1).hash as usize)), _MM_HINT_T0);
+
+                            let feature = fb.ffm_buffer.get_unchecked(ffm_buffer_index);
+                            let feature_value = feature.value as f32;
+
+                            let mut feature_index = feature.hash as usize;
+                            let mut offset: usize = field_index_ffmk as usize;
+
+                            if feature_num == 0 {
+                                for _z in 0..ffm_fields_count_as_usize {
+                                    _mm_prefetch(mem::transmute::<&f32, &i8>(&ffm_weights.get_unchecked(feature_index + ffmk_as_usize)), _MM_HINT_T0);
+                                    for k in 0..ffmk_as_usize {
+                                        *contra_fields.get_unchecked_mut(offset + k) = ffm_weights.get_unchecked(feature_index + k) * feature_value;
                                     }
-                                });
-                                ffm_buffer_index += 1;
-                                feature_num += 1;
+
+                                    offset += fc;
+                                    feature_index += ffmk_as_usize;
+                                }
+                            } else {
+                                for _z in 0..ffm_fields_count_as_usize {
+                                    _mm_prefetch(mem::transmute::<&f32, &i8>(&ffm_weights.get_unchecked(feature_index + ffmk_as_usize)), _MM_HINT_T0);
+                                    for k in 0..ffmk_as_usize {
+                                        *contra_fields.get_unchecked_mut(offset + k) += ffm_weights.get_unchecked(feature_index + k) * feature_value;
+                                    }
+
+                                    offset += fc;
+                                    feature_index += ffmk_as_usize;
+                                }
                             }
+
+                            ffm_buffer_index += 1;
+                            feature_num += 1;
                         }
+                    }
+
+                    let mut ffm_values_offset = 0;
+                    for (i, feature) in fb.ffm_buffer.iter().enumerate() {
+                        let feature_value = feature.value;
+                        let feature_index = feature.hash as usize;
+                        let feature_contra_field_index = feature.contra_field_index as usize;
+
```

Comment:
Not needed spaces

Reply:
Tbh I actually want to reach a point where we just do `cargo fmt`.